token/program/src/nostd_adapter.rs

```diff
@@ -0,0 +1,118 @@
+use std::ops::DerefMut;
+
+use solana_nostd_entrypoint::NoStdAccountInfo;
+use solana_program::{
+    account_info::AccountInfo as SolanaAccountInfo, program_error::ProgramError, pubkey::Pubkey,
+};
+
+pub trait IAccountInfo {
+    fn lamports(&self) -> u64;
+    fn key(&self) -> &Pubkey;
+    fn owner(&self) -> &Pubkey;
+    fn is_signer(&self) -> bool;
+    fn is_writable(&self) -> bool;
+    fn data_len(&self) -> usize;
+    fn reassign(&self, new_owner: &Pubkey);
+    fn realloc(&self, new_len: usize, is_mutable: bool) -> Result<(), ProgramError>;
+    unsafe fn borrow_data(&self) -> &[u8];
+    unsafe fn borrow_data_mut(&self) -> &mut [u8];
+    unsafe fn borrow_mut_lamports(&self) -> &mut u64;
+}
+
+impl IAccountInfo for NoStdAccountInfo {
+    fn lamports(&self) -> u64 {
+        unsafe { *self.unchecked_borrow_lamports() }
+    }
+
+    fn key(&self) -> &Pubkey {
+        NoStdAccountInfo::key(self)
+    }
+
+    fn owner(&self) -> &Pubkey {
+        NoStdAccountInfo::owner(self)
+    }
+
+    fn is_signer(&self) -> bool {
+        NoStdAccountInfo::is_signer(self)
+    }
+
+    fn is_writable(&self) -> bool {
+        NoStdAccountInfo::is_writable(self)
+    }
+
+    fn data_len(&self) -> usize {
+        NoStdAccountInfo::data_len(self)
+    }
+
+    unsafe fn borrow_mut_lamports(&self) -> &mut u64 {
+        unsafe { NoStdAccountInfo::unchecked_borrow_mut_lamports(self) }
+    }
+
+    unsafe fn borrow_data(&self) -> &[u8] {
+        unsafe { NoStdAccountInfo::unchecked_borrow_data(self) }
+    }
+
+    unsafe fn borrow_data_mut(&self) -> &mut [u8] {
+        unsafe { NoStdAccountInfo::unchecked_borrow_mut_data(self) }
+    }
+
+    fn reassign(&self, new_owner: &Pubkey) {
+        NoStdAccountInfo::reassign(self, new_owner)
+    }
+
+    fn realloc(&self, new_len: usize, is_mutable: bool) -> Result<(), ProgramError> {
+        NoStdAccountInfo::realloc(self, new_len, is_mutable)
+    }
+}
+
+impl<'a> IAccountInfo for SolanaAccountInfo<'a> {
+    fn lamports(&self) -> u64 {
+        SolanaAccountInfo::lamports(self)
+    }
+
+    fn key(&self) -> &Pubkey {
+        &self.key
+    }
+
+    fn owner(&self) -> &Pubkey {
+        &self.owner
+    }
+
+    fn is_signer(&self) -> bool {
+        self.is_signer
+    }
+
+    fn is_writable(&self) -> bool {
+        self.is_writable
+    }
+
+    fn data_len(&self) -> usize {
+        self.data.borrow().len()
+    }
+
+    fn reassign(&self, new_owner: &Pubkey) {
+        SolanaAccountInfo::assign(self, new_owner);
+    }
+
+    fn realloc(&self, new_len: usize, is_mutable: bool) -> Result<(), ProgramError> {
+        SolanaAccountInfo::realloc(self, new_len, is_mutable)
+    }
+
+    unsafe fn borrow_data(&self) -> &[u8] {
```

Comment:
I'm not an expert at unsafe rust: can we make this function safe but add the unsafe block into the impl? That way we don't need to use `unsafe` everywhere in processor.rs

Reply:
I guess the reason to not do this is code readability? Explicity having the unsafe in processor makes it easier to understand that the actions taken are actually memory safe

Reply:
yeah we can definitely do that, it depends on how much we wanna leak the "unsafe" semantic to the outside of the world.

Since it's only used internally i think it makes sense to keep the unsafe block inside the impl